src/Aspire.Dashboard/Model/ResourceOutgoingPeerResolver.cs

```diff
@@ -0,0 +1,87 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Aspire.Dashboard.Model;
+
+public sealed class ResourceOutgoingPeerResolver : IOutgoingPeerResolver, IAsyncDisposable
+{
+    private readonly IDashboardViewModelService _dashboardViewModelService;
+    private readonly Dictionary<string, ResourceViewModel> _resourceNameMapping = new();
+    private readonly CancellationTokenSource _watchContainersTokenSource = new();
+    private readonly Task _watchTask;
+    private readonly object _lock = new object();
+
+    public ResourceOutgoingPeerResolver(IDashboardViewModelService dashboardViewModelService)
+    {
+        _dashboardViewModelService = dashboardViewModelService;
+
+        var viewModelMonitor = _dashboardViewModelService.GetResources();
+        var initialList = viewModelMonitor.Snapshot;
+        var watch = viewModelMonitor.Watch;
+
+        foreach (var result in initialList)
+        {
+            _resourceNameMapping[result.Name] = result;
+        }
+
+        _watchTask = Task.Run(async () =>
```

Comment:
Given that the IDashboardViewModelService is already caching the changes (at least for the lifetime of that object, and we need to resolve your other comment on that matter first I suppose), I'm wondering if we need this whole extra layer of caching on top of it. I guess the call to GetResourceMonitor (inside GetResources) is too expensive to do repeatedly. Just seems like a bit much for what amounts to "iterate the current snapshot". I wonder if we could expose the current snapshot better somehow instead? Just spitballing. Nothing wrong with this class as it is.

Reply:
I copied what the console logs page does.

I think `DashboardViewModelService` is very complex. A method to subscribe to resource updates and a method to get the resources would be less efficient (does it matter?) but much simpler.

Reply:
Yeah the console logs needs the watcher part to be able to update the list as it changes. But theoretically here you only need the current snapshot at the time of a lookup. I think we can tweak the view model service api to make it simpler. We've added/removed various methods of accessing as we've needed them/not needed them anymore. At minimum I think we could add a call to get the current snapshot. I guess for your use case though, when would you call that? On every lookup? It could theoretically get stale otherwise... but would that matter in practice (especially since this is not the page you hit when the dashboard loads)?

Reply:
> I guess for your use case though, when would you call that? On every lookup? It could theoretically get stale otherwise... but would that matter in practice (especially since this is not the page you hit when the dashboard loads)?

That's why there is a callback to tell you data has changed and a method to get the data. That's the pattern telemetry uses.

Debounce logic to throttle callbacks could be used if there are many changes. e.g. wait 100ms between callback notifications. (I haven't done that in telemetry yet, but I plan to)